### DISCUSSION TITLE
Sync BlackHole clock with other devices

### DISCUSSION THREAD

**Discussion by xavrr** [ExistentialAudio/BlackHole#27] [Feature Requests]
Hello, I'm using blackhole on catalina. (Many thanks for this great work !)

I'm using blackhole to record audio during many hours and I've noticed that it seems to introduce a delay that will add up to multiple seconds after couple of hours.
Changing the audio frequency and back using audio midi setup seems to "reset" the latency.

Does anyone have experienced the same issue ?

**Comment by devinroth** [comment]
First off BlackHole doesn’t currently support syncing clocks with other devices. After a while you’ll notice a delay as the clock get further out of sync. What programs are you using with BlackHole? If you are using a multi output or aggregate make sure drift correction is enabled. If the audio software you are using does support drift correction this will unfortunately continue. 

Clock syncing is a very important issue for me. My goal is to allow users to select an audio device to use as a clock source.

**Comment by xavrr** [comment]
Thanks for your answer.
My setup is  python midi event generator>garageband -> blackhole -> python audio recorder.
I'm recording both midi events and audio, and I really need it to be synced.
Following your message, II'm trying to use an aggregated device (mac microphone+blackhole) using the mac microphone as clock source and checking drift correction on blackhole. 
Does it have a chance of making a change ?

**Comment by devinroth** [comment]
It should work in theory with drift correction but aggregate devices have been known to be sketchy. Ultimately clock syncing is the solution but I haven’t had time to dig into that yet. Let me know the result.

**Comment by ivanvfer** [comment]
Hi, first of all many thanks for this cool development! 

I use Tonebridge to support different guitar effects, and with BlackHole I can route the guitar sound to Reaper (using BlackHole channels other than 1-2). With Reaper, I can mix it with the mic input, and route the mixed audio to Skype using the first two channels of BlackHole.

I have also noticed that, after playing for some time, there was a small delay in the guitar sound. I used an aggregate device in Reaper, with BlackHole and a sound card. Enabling drift correction, with the sound card as the clock source, seems to have addressed the delay issue (at least, guitar delay is not noticeable after ~2h).

**Comment by HEnquist** [comment]
It would be great if it was possible to either use another device as clock master, or to expose some kind of control for fine adjusment of the virtual samplerate of blackhole.
My own use case is to set blackhole as the default output device. Then I record the output from blackhole, run it through my DSP software, and play it back on another device. If I could lock the blackhole clock to the real output device, I wouldn't need to use asynchronous resampling to keep the two in sync.

**Comment by devinroth** [comment]
It’s in development. But I think I’m going to save this feature for the pro version I’m working on.

Initial tests work well but there are some bugs I need to iron out.

Devin Roth

**Reply by HEnquist** [reply]
How will this work, do you just select another device as clock source? Do you dare to guess when it will be ready?
And does Pro mean a paid closed source version?

**Comment by devinroth** [comment]
Yeah simply select a different device as the clock.

And yes I plan on monetizing some advanced features.

Devin Roth

**Comment by audiogamble** [comment]
I noticed a branch that included this feature when I was pulling the repository from xcode. Is there anything I need to do to get that working, or is the feature still a work in progress?

**Comment by devinroth** [comment]
Sorry that branch does not work.

Devin Roth

**Comment by niugm** [comment]
Hi,  have any progress to share? I really need this feature

**Comment by Bay40k** [comment]
Chiming in to say I'm _still_ having this exact issue, using an aggregate audio device

**Comment by Mitchco** [comment]
Hi Devin, thanks again for BlackHole, great work! Just enquiring about this feature. No problem paying for a Pro version...

Kind regards,
Mitch

**Comment by devinroth** [comment]
It’s been on the back burner for a bit. Still part of the plan but I haven’t had a chance to finish.

Devin

**Reply by David8266** [reply]
Hi Devin. Thanks for developing this software. It is essential for me. The clock synchronization feature with output device would be really great and would avoid the ASRC. it is true that it is completely digital so there is no jitter problem, but with 16 ASRC channels (my case) the CPU works a lot and penalizes the performance of the DAW.
If this feature is released in the paid version, I will be a buyer!

**Comment by Mark8869** [comment]
Any timeline for this feature?
It'll be truly game changer

**Reply by devinroth** [reply]
Sorry no timeline.

**Comment by MuYanQin** [comment]
Hello. May I ask when there is a plan to launch this clock Source select version?

**Comment by devinroth** [comment]
It’s on the back burner for now. Sorry.

Devin
